designs/0035-cmdstanrs.md

```diff
@@ -0,0 +1,348 @@
+judgment is that they are useful, but are best developed separately.
+
+### Serialization
+
+It is trivial to provide a function such as 
+```rust
+fn write_json<T: Serialize>(data: &T, file: &Path) {}
+```
+but it serves no purpose -- it does not enforce the conventions
+adopted for representing Stan data types in JSON (e.g. matrices
+represented as a vector of *row* vectors, not a vector of *column*
+vectors), hence, would likely lead to unexpected (and potentially
+silent!) errors.
+
+In order to develop a serializer which respects the conventions for
+Stan data types, one would need to declare conventions for the mapping
+of Rust data types to Stan data
+types. [serde_json](https://github.com/serde-rs/json) would be nice to
+use, but has some incompatibilities (Rust tuple is represented as an
+array, rather than an object).
+
+Moreover, to represent matrices, complex numbers, etc., one would need
+to support types from the crate ecosystem since the standard library
+lacks these -- [nalgebra](https://github.com/dimforge/nalgebra) and
+[num-complex](https://github.com/rust-num/num-complex) are reasonable
+choices, but nonetheless represent decisions to be made!
+
+From a design perspective, this is a great place to defer to the user,
+at least for the moment. A principled approach would involve writing a
+data format for [serde](https://serde.rs/).
+
+### Deserialization
+
+Parsing Stan CSVs to a strongly-typed representation is simple if one
+wishes to simply obtain a matrix of values (or `Vec<Vec<f64>>` if we
+limit ourselves to `std` library types). However, one needs to extract
+the variables from each line, thus, one needs to know the types (and
+their dimensions). A recursive definition of types using `enum`s and
+`struct`s could probably work to represent such a thing in Rust, but
+may not necessarily be particularly ergonomic (i.e. much unavoidable
+boilerplate would be needed to use the resultant type).
+
+Procedural macros, applied to a Stan program stored in a string
+literal in a Rust program, could be used to generate types and a
+parser for Stan CSVs produced said program. However, in order to
+implement such an idea, one would first need to adopt conventions for
+representing Stan data types using Rust data types. This requires
+careful thought and and is something best left for the future, if
+ever.
+
+The current proposal is for `CmdStanOutput` to be capable of returning
+paths to the files and the user parses them however they desire.
+This leaves open the possibility of multiple parsing strategies.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Representing CmdStan arguments/options as a concrete syntax tree is
+potentially brittle. If the CmdStan grammar undergoes radical change,
+this interface will need to change accordingly. However, the CmdStan
+grammar is intended to be quite stable. Moreover, it is not
+necessarily the case that radical changes to the CmdStan grammar could
+be hidden behind something other abstraction.
```

Comment:
Since breaking changes are discussed here, it seems like a reasonable place to ask about versioning: I assume the goal is to emulate cmdstanpy/r and be "mostly" cmdstan-version-independent? Or would a given version of this crate be specialized for a specific cmdstan version?

Version independence is nice in theory but IMO has some practical concerns which may lead one to reconsider the decisions make when working on cmdstanpy/r, especially when it comes to static checking. If cmdstan adds a new argument, supporting older versions simultaneously now requires the builder to do some runtime checks on top of what the Rust type system provided you already, which is a bummer. 

Alternatively, the argument structs could themselves be versioned... but that seems like a pain.

Reply:
Technically, quite a few (maybe all?) of the structs and enums would need to be marked with `#[non_exhaustive]`; I was anticipating as much with the builder pattern.

If the Rust code stays up-to-date with the latest cmdstan arguments and options, then the worst case is that an older cmdstan will return something to the effect of "... is not a valid value for ..." in the stderr. An output is already not a guarantee (the process could die for any reason), thus, rather than leaving it to the user, we could provide a course classification of the errors (insufficient cmdstan version to support the user's argument tree being a variant).

I do not see an easy way to raise cmdstan capability checks to compile-time without resorting to strange things in a build script or procedural macro antics. Even then, the cmdstan installation that a user intends to use must be available at compile time.

Alternatively, one could package cmdstan as a Rust crate and manage it as a dependency with cargo. I do not recommend this due to the other headaches it would create.

Reply:
That seems reasonable! Saying as much in the document would be appreciated